Proofs/Experiments/MemoryAliasing.lean

```diff
@@ -8,22 +8,586 @@ The goal is to eliminate the sorry, and to simplify the proof to a tactic invoca
+      by_cases h₂ : i - 8 < n' * 8
+      · simp [h₂]
+        simp [show (i < n' * 8 + 8) by omega]
+        have hi' : ∃ i', i = i' + 8 := by
+          apply Classical.byContradiction
+          intros h
+          simp at h
+          specialize h (i - 8)
+          omega
+        obtain ⟨i', hi'⟩ := hi'
+        subst hi'
+        simp
+        congr 2
+        rw [BitVec.add_assoc]
+        congr
+        rw [BitVec.add_def]
+        congr 1
+        simp
+        rw [Nat.mod_eq_of_lt]
+        · omega
+        · omega
+      · simp [h₂]
+        intros h₃
+        omega
+      · omega
+
+/-- info: 'getLsb_read_mem_bytes'' depends on axioms: [propext, Classical.choice, Quot.sound] -/
+#guard_msgs in #print axioms getLsb_read_mem_bytes'
+
+-- (FIXME) Make write_mem private.
+-- We export write_mem_bytes, not write_mem.
+def write_mem' (addr : BitVec 64) (val : BitVec 8) (s : Memory) : Memory :=
+  write_store addr val s
+
+theorem write_mem'_of_eq (hix : ix = addr) : write_mem' addr val s ix = val := by
+  simp only [write_mem']
+  subst ix
+  apply store_read_over_write_same
+
+theorem write_mem'_of_neq (hix : ix ≠ addr) : write_mem' addr val s ix = s ix := by
+  simp only [write_mem']
+  apply store_read_over_write_different
+  assumption
+
+@[simp]
+theorem write_mem_eq_write_mem' :  (write_mem addr val s).mem = write_mem' addr val s.mem := rfl
+
+def write_mem_bytes' (n : Nat) (addr : BitVec 64) (val : BitVec (n * 8)) (s : Memory) : Memory :=
+  match n with
+  | 0 => s
+  | n' + 1 =>
+    let byte := BitVec.extractLsb 7 0 val
+    let s := write_mem' addr byte s
+    let val_rest := BitVec.zeroExtend (n' * 8) (val >>> 8)
+    write_mem_bytes' n' (addr + 1#64) val_rest s
+
+/-- Defintion to extract the `n`th least significant *Byte* from a bitvector. -/
+def BitVec.extractLsB (val : BitVec w₁) (n : Nat) : BitVec 8 :=
+  val.extractLsb ((n + 1)*8 - 1) (n * 8) |>.cast (by omega)
+
+theorem BitVec.extractLsB_def (val : BitVec w₁) (n : Nat):
+    val.extractLsB n = (val.extractLsb ((n + 1)*8 - 1) (n * 8) |>.cast (by omega)) := rfl
+
+@[simp]
+theorem BitVec.getLsb_extractLsB (val : BitVec w₁) :
+    ((BitVec.extractLsB val n).getLsb i) =
+    (decide (i ≤ 7) && val.getLsb (n * 8 + i)) := by
+  simp [extractLsB]
+  rw [Nat.succ_mul]
+  simp only [Nat.add_one_sub_one]
+  simp [Nat.add_sub_cancel_left]
+
+/-- Writing zero bytes does not change memory. -/
+theorem write_mem_bytes'_zero : write_mem_bytes' 0 addr val s = s := rfl
+
+/-- Writiing (n + 1) bytes can be described as writing `n` bytes, and then recursing to write the rest. -/
+theorem write_mem_bytes'_succ :
+    write_mem_bytes' (n + 1) addr val s =
+    let byte := BitVec.extractLsb 7 0 val
+    let s := write_mem' addr byte s
+    let val_rest := BitVec.zeroExtend (n * 8) (val >>> 8) -- TODO: rewrite this as 'truncate'.
+    write_mem_bytes' n (addr + 1#64) val_rest s := rfl
+
+def write_mem_bytes'₂ (n : Nat) (addr : BitVec 64) (val : BitVec (n * 8)) (s : Memory) : Memory :=
```

Comment:
Subcripts like `₂` in a function name are very new to me.

Reply:
recall that this code is experimental, and exposes multiple definitions which we are experimenting with.